README
This package contains a few python modules that used to live in mate-python-extras:

  - mateapplet
  - matewnck
  - cajaburn
  - marco

Note: this branch requires a special branch of WAF: https://code.launchpad.net/~gjc/waf/cmd
This WAF version can be retrieved by this command:

$ bzr co lp:~gjc/waf/cmd waf-cmd



NEWS
1.5.0

 * Removed totem
 * Removed mediaprofiles


tests/common.py
import os
import sys
import gobject
gobject.threads_init()


modules = [
    "mateapplet",
    "matewnck",
    "cajaburn",
    "marco",
    "matekeyring",
    "matedesktop",
    "atril",
    ]

def run_import_tests(builddir, no_import_hooks):
    if not no_import_hooks:
        import ltihooks
        ltihooks.install()

    for item in modules:
        if isinstance(item, tuple):
            module, dirname = item
        else:
            module = item
            dirname = item
        sys.path.insert(0, os.path.join(builddir, dirname))
        print "Trying to import module %s... " % (module,),
        try:
            mod = __import__(module) # try to import the module to catch undefined symbols
        except ImportError, ex:
            if ex.args[0].startswith("No module named"):
                print "not found"
            else:
                raise
        else:
            print "ok."
            globals()[module] = mod

    if not no_import_hooks:
        ltihooks.uninstall()


tests/ltihooks.py
# -*- Mode: Python; py-indent-offset: 4 -*-
# ltihooks.py: python import hooks that understand libtool libraries.
# Copyright (C) 2000 James Henstridge.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

import os, ihooks

class LibtoolHooks(ihooks.Hooks):
    def get_suffixes(self):
        """Like normal get_suffixes, but adds .la suffixes to list"""
	ret = ihooks.Hooks.get_suffixes(self)
	ret.insert(0, ('module.la', 'rb', 3))
	ret.insert(0, ('.la', 'rb', 3))
	return ret

    def load_dynamic(self, name, filename, file=None):
        """Like normal load_dynamic, but treat .la files specially"""
	if len(filename) > 3 and filename[-3:] == '.la':
	    fp = open(filename, 'r')
	    dlname = ''
	    installed = 1
	    line = fp.readline()
	    while line:
		if len(line) > 7 and line[:7] == 'dlname=':
		    dlname = line[8:-2]
		elif len(line) > 10 and line[:10] == 'installed=':
		    installed = line[10:-1] == 'yes'
		line = fp.readline()
	    fp.close()
	    if dlname:
		if installed:
		    filename = os.path.join(os.path.dirname(filename),
					    dlname)
		else:
		    filename = os.path.join(os.path.dirname(filename),
					    '.libs', dlname)
	return ihooks.Hooks.load_dynamic(self, name, filename, file)

importer = ihooks.ModuleImporter()
importer.set_hooks(LibtoolHooks())

def install():
    importer.install()
def uninstall():
    importer.uninstall()

install()


autogen.sh
#!/bin/sh
# Run this to generate all the initial makefiles, etc.

srcdir=`dirname $0`
test -z "$srcdir" && srcdir=.

PKG_NAME="mate-python-desktop"
REQUIRED_AUTOCONF_VERSION=2.59
REQUIRED_AUTOMAKE_VERSION=1.11

(test -f $srcdir/configure.ac \
  && test -f $srcdir/autogen.sh) || {
    echo -n "**Error**: Directory "\`$srcdir\'" does not look like the"
    echo " top-level $PKG_NAME directory"
    exit 1
}

./generate-ChangeLog

DIE=0

# This is a bit complicated here since we can't use mate-config yet.
# It'll be easier after switching to pkg-config since we can then
# use pkg-config to find the mate-autogen.sh script.

mate_autogen=
mate_datadir=

ifs_save="$IFS"; IFS=":"
for dir in $PATH ; do
  test -z "$dir" && dir=.
  if test -f $dir/mate-autogen ; then
    mate_autogen="$dir/mate-autogen"
    mate_datadir=`echo $dir | sed -e 's,/bin$,/share,'`
    break
  fi
done
IFS="$ifs_save"

if test -z "$mate_autogen" ; then
  echo "You need to install the mate-common module and make"
  echo "sure the mate-autogen.sh script is in your \$PATH."
  exit 1
fi

MATE_DATADIR="$mate_datadir" USE_MATE2_MACROS=1 . $mate_autogen


atril/atrilmodule.c
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

/* include this first, before NO_IMPORT_PYGOBJECT is defined */
#include <pygobject.h>
#include <pygtk/pygtk.h>

#include <atril-document.h>
#include <pycairo/pycairo.h>

Pycairo_CAPI_t *Pycairo_CAPI;


void pyatril_register_classes (PyObject *d);
void pyatril_add_constants(PyObject *module, const gchar *strip_prefix);

extern PyMethodDef pyatril_functions[];

PyObject *
_wrap_ev_shutdown (void)
{
/*     g_message("ev_shutdown();"); */
    ev_shutdown();
    Py_INCREF(Py_None);
    return Py_None;
}


DL_EXPORT(void)
initatril(void)
{
    PyObject *m, *d;

    /* Init glib threads asap */
    if (!g_thread_supported ())
       g_thread_init (NULL);

    init_pygobject ();

    pyg_enable_threads ();

    Pycairo_IMPORT;

    ev_init ();

    m = Py_InitModule ("atril", pyatril_functions);
    d = PyModule_GetDict (m);

    pyatril_register_classes (d);
    pyatril_add_constants(m, "EV_");

    if (PyErr_Occurred ()) {
        return;
    }

      /* Call ev_shutdown() on an atexit handler (bug #570622) */
    {
        // note: py_atexit_method_def has to be static, since python keeps a pointer to it
        static PyMethodDef py_atexit_method_def = {NULL, (PyCFunction)_wrap_ev_shutdown, METH_NOARGS, NULL};
        PyObject *py_atexit_func = PyCFunction_NewEx(&py_atexit_method_def, NULL, NULL);
        PyObject *atexit = PyImport_ImportModule("atexit");
        PyObject_CallMethod(atexit, "register", "N", py_atexit_func);
        Py_DECREF(atexit);
    }
}


bzrplugins/gnulog.py
# Copyright (C) 2006-2008 Gustavo J. A. M. Carneiro  <gjcarneiro@gmail.com>

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

"""
A bzr plugin that allows bzr log to output in GNU ChangeLog format.

Installation:
  Copy this file to ~/.bazaar/plugins/

Usage:
  bzr log -v --log-format 'gnu'

Changes:
  - 2008-12-08: include information of fixed bugs (bzr commit --fixes)
"""

import time
import textwrap
from cStringIO import StringIO
import codecs

Encoder = codecs.getwriter('ascii')

import bzrlib.log
#from bzrlib.osutils import format_date

class GnuFormatter(bzrlib.log.LogFormatter):

    def __init__(self, *args, **kwargs):
        bzrlib.log.LogFormatter.__init__(self, *args, **kwargs)
        self._changes_buffer = StringIO()
        self._date_line = None
        self.supports_tags = True
        self.supports_delta = True

    def log_revision(self, lr):
        self.show(lr.revno, lr.rev, lr.delta, lr.tags)
    
    def show(self, revno, rev, delta, tags=None):
        to_file = Encoder(self.to_file, errors='replace')
        if tags is not None:
            self._flush()
            print >> to_file, "=== %s ===" % ', '.join(tags)
        date_str = time.strftime("%Y-%m-%d", time.localtime(rev.timestamp))
        author = rev.get_apparent_author().strip()
        date_line = date_str + "  " + author
        if date_line != self._date_line:
            self._flush()
            self._date_line = date_line
        self._show_changes(revno, rev, delta)

    def _flush(self):
        if self._date_line is None or not self._changes_buffer.getvalue():
            return
        to_file = Encoder(self.to_file, errors='replace')
        print >> to_file, self._date_line
        print >> to_file
        to_file.write(self._changes_buffer.getvalue())
        self._changes_buffer.truncate(0)

    def begin_log(self):
        pass

    def end_log(self):
        self._flush()

    def __del__(self): # this is evil, I need bzr to call a method on me when it's finished..
        self._flush()

    def _show_changes(self, revno, rev, delta):
        to_file = Encoder(self._changes_buffer, errors='replace')
        output_standalone_message = True

        message = "[%s] %s" % (revno, rev.message)
        bugs_text = rev.properties.get('bugs', '')
        if bugs_text:
            if message[-1] in '.;':
                message = message[:-1]
            message = "%s; %s" % (message, bugs_text)
        
        if delta is not None:
            ## special case when only text modifications exist
            if delta.modified and not (delta.added or delta.removed
                                       or delta.renamed):
                for num, (path, id, kind, text_modified, meta_modified) \
                        in enumerate(delta.modified):
                    if num == len(delta.modified) - 1:
                        for line in textwrap.wrap(
                                "%s: %s" % (path, message),
                                width=70,
                                initial_indent="* ",
                                subsequent_indent=""):
                            print >> to_file, "\t" + line
                    else:
                        print >> to_file, "\t* %s," % (path,)
                output_standalone_message = False
                print >> to_file

            else:

                self._flush()
                
                if delta.added:
                    for num, (path, id, kind) in enumerate(delta.added):
                        if num == len(delta.added) - 1:
                            print >> to_file, "\t* %s: Added." % (path,)
                            print >> to_file
                        else:
                            print >> to_file, "\t* %s," % (path,)

                if delta.removed:
                    for num, (path, id, kind) in enumerate(delta.removed):
                        if num == len(delta.removed) - 1:
                            print >> to_file, "\t* %s: Removed." % (path,)
                            print >> to_file
                        else:
                            print >> to_file, "\t* %s," % (path,)

                if delta.modified:
                    for num, (path, id, kind, text_modified, meta_modified) \
                            in enumerate(delta.modified):
                        if num == len(delta.modified) - 1:
                            print >> to_file, "\t* %s: Modified." % (path,)
                            print >> to_file
                        else:
                            print >> to_file, "\t* %s," % (path,)

                if delta.renamed:
                    for (oldpath, newpath, id, kind,
                         text_modified, meta_modified) in delta.renamed:
                        if text_modified:
                            line = ("%s: Renamed to %s and modified."\
                                    % (oldpath, newpath))
                        else:
                            line = "%s: Renamed to %s." % (oldpath, newpath)
                        for l1 in textwrap.wrap(line, 70,
                                                initial_indent="* ",
                                                subsequent_indent="  "):
                            print >> to_file, "\t" + l1
                    print >> to_file

        if output_standalone_message:
            ## The log message, on a line by itself
            if not message:
                print >> to_file,  '\t(no message)'
            else:
                for l2 in textwrap.wrap(message, 70):
                    print >> to_file,  '\t' + l2

            print >> to_file
            self._flush()


bzrlib.log.register_formatter('gnu', GnuFormatter)



cajaburn/nb_module.c
/* 
 * Copyright (C) 2004 Tiago Cogumbreiro <cogumbreiro@users.sf.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.
 *
 * Authors: Tiago Cogumbreiro <cogumbreiro@users.sf.net>
 */

#include <Python.h>
#include <pygobject.h>
#include "nb_track.h"
#include <caja-burn-recorder.h>
void caja_burn_register_classes (PyObject *d);
void caja_burn_add_constants(PyObject *module, const gchar *strip_prefix);
extern PyMethodDef caja_burn_functions[];



DL_EXPORT(void)
initcajaburn(void)
{
	PyObject *m, *d;
	
	init_pygobject ();
	
	m = Py_InitModule ("cajaburn", caja_burn_functions);
	d = PyModule_GetDict (m);
	
	/* init Track classes */
	if (nb_track_init (m))
		return;
	/* init auto-gened classes */
	caja_burn_register_classes (d);
	caja_burn_add_constants (m, "CAJA_BURN_");
}


examples/wnck_example.py
import gtk
import matewnck

def main():
    screen = matewnck.screen_get_default()
    # Process pending gtk+ events so that matewnck can find out about existing windows.
    while gtk.events_pending():
        gtk.main_iteration()
    for window in screen.get_windows():
        # A XID is a number that identifies a X window.
        xid = window.get_xid()
        # By calling gtk.gdk.window_foreign_new(xid), we are creating
        # a gtk.gdk.Window which wraps the window that is identified by
        # the XID. It will return None if it can't find the X window
        # identified by the XID in question.
        wrapped_window = gtk.gdk.window_foreign_new(xid)
        # We are maximizing the window here, but
        # you can do almost anything with a gtk.gdk.Window.
        wrapped_window.maximize()

if __name__ == '__main__':
    main()


cajaburn/nb_track.c
/* 
 * Copyright (C) 2004 Tiago Cogumbreiro <cogumbreiro@users.sf.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.
 *
 * Authors: Tiago Cogumbreiro <cogumbreiro@users.sf.net>
 */
#include <Python.h>
#include <caja-burn-recorder.h>
#include "nb_track.h"

/* base class */

static void
nb_Track_dealloc(nb_Track* self)
{
	if (self->track.type == CAJA_BURN_RECORDER_TRACK_TYPE_DATA) {
		g_free (self->track.contents.data.filename);
	} else if (self->track.type == CAJA_BURN_RECORDER_TRACK_TYPE_AUDIO) {
		g_free (self->track.contents.audio.filename);
		g_free (self->track.contents.audio.cdtext);
	}
    self->ob_type->tp_free((PyObject*)self);
}

static PyObject *
nb_Track_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
{
    nb_Track *self;
	
    self = (nb_Track *)type->tp_alloc(type, 0);
    if (self != NULL) {
	    /* make sure it points to an empty string */
	    self->track.contents.audio.filename = NULL;
	    self->track.contents.audio.cdtext = NULL;
    }

    return (PyObject *)self;
}

/* setters */

static int
nb_AudioTrack_setfilename (nb_Track *self, PyObject *value, void *closure)
{
	if (value == NULL) {
		PyErr_SetString(PyExc_TypeError, "Cannot delete the filename attribute");
		return -1;
	}

	if (! PyString_Check(value)) {
		PyErr_SetString(PyExc_TypeError, 
                        "The filename attribute value must be a string");
		return -1;
	}
	g_free (self->track.contents.audio.filename);
	self->track.contents.audio.filename = g_strdup (PyString_AsString (value));
	return 0;
}

static int
nb_AudioTrack_setcdtext (nb_Track *self, PyObject *value, void *closure)
{
	if (value == NULL) {
		PyErr_SetString(PyExc_TypeError, "Cannot delete the cdtext attribute");
		return -1;
	}

	if (! PyString_Check(value)) {
		PyErr_SetString(PyExc_TypeError, 
                        "The cdtext attribute value must be a string");
		return -1;
	}
	g_free (self->track.contents.audio.cdtext);
	self->track.contents.audio.cdtext = g_strdup (PyString_AsString (value));
	return 0;
}

static int
nb_DataTrack_setfilename (nb_Track *self, PyObject *value, void *closure)
{
	if (value == NULL) {
		PyErr_SetString(PyExc_TypeError, "Cannot delete the filename attribute");
		return -1;
	}

	if (! PyString_Check(value)) {
		PyErr_SetString(PyExc_TypeError, 
                        "The filename attribute value must be a string");
		return -1;
	}
	g_free (self->track.contents.data.filename);
	self->track.contents.data.filename = g_strdup (PyString_AsString (value));
	return 0;
}

/* getters */

static PyObject *
nb_AudioTrack_getfilename(nb_Track *self, void *closure)
{
    return Py_BuildValue ("s", self->track.contents.audio.filename);
}

static PyObject *
nb_AudioTrack_getcdtext(nb_Track *self, void *closure)
{
    return Py_BuildValue ("s", self->track.contents.audio.cdtext);
}

static PyObject *
nb_DataTrack_getfilename(nb_Track *self, void *closure)
{
    return Py_BuildValue ("s", self->track.contents.data.filename);
}

/* initializers */
static int
nb_AudioTrack_init(nb_Track *self, PyObject *args, PyObject *kwds)
{
	char *filename=NULL, *cdtext=NULL;

	static char *kwlist[] = {"filename", "cdtext", NULL};

	if (! PyArg_ParseTupleAndKeywords(args, kwds, "|ss", kwlist, 
                                      &filename, &cdtext))
		return -1; 

	if (!filename)
		filename = "";
    
	if (!cdtext)
		cdtext = "";
    	
    self->track.type = CAJA_BURN_RECORDER_TRACK_TYPE_AUDIO;
    self->track.contents.audio.filename = g_strdup (filename);
    self->track.contents.audio.cdtext = g_strdup (cdtext);
	return 0;
}

static int
nb_DataTrack_init(nb_Track *self, PyObject *args, PyObject *kwds)
{
	char *filename=NULL;

	static char *kwlist[] = {"filename", NULL};

	if (! PyArg_ParseTupleAndKeywords(args, kwds, "|s", kwlist,  &filename))
		return -1; 

	if (!filename)
		filename = "";
    
    self->track.type = CAJA_BURN_RECORDER_TRACK_TYPE_DATA;
    self->track.contents.data.filename = g_strdup (filename);
	return 0;
}

static PyTypeObject nb_Track_Type = {
    PyObject_HEAD_INIT(NULL)
    0,                         /*ob_size*/
    "cajaburn.Track",             /*tp_name*/
    sizeof(nb_Track),             /*tp_basicsize*/
    0,                         /*tp_itemsize*/
    0, /*tp_dealloc*/
    0,                         /*tp_print*/
    0,                         /*tp_getattr*/
    0,                         /*tp_setattr*/
    0,                         /*tp_compare*/
    0,                         /*tp_repr*/
    0,                         /*tp_as_number*/
    0,                         /*tp_as_sequence*/
    0,                         /*tp_as_mapping*/
    0,                         /*tp_hash */
    0,                         /*tp_call*/
    0,                         /*tp_str*/
    0,                         /*tp_getattro*/
    0,                         /*tp_setattro*/
    0,                         /*tp_as_buffer*/
    Py_TPFLAGS_DEFAULT, /*tp_flags*/
    "Represents a recorder's track.",           /* tp_doc */
};

/* AudioTrack definition */
static PyGetSetDef nb_AudioTrack_getseters[] = {
    {"filename",
     (getter)nb_AudioTrack_getfilename, (setter)nb_AudioTrack_setfilename,
     "The track filename, if it's a WAV file must have .wav extension. "
     "Otherwise it will be considered a RAW file.",
     NULL},
    {"cdtext", 
     (getter)nb_AudioTrack_getcdtext, (setter)nb_AudioTrack_setcdtext,
     "The track's associated cdtext. Currently it's not implemented.",
     NULL},
    {NULL}  /* Sentinel */
};

static PyGetSetDef nb_DataTrack_getseters[] = {
    {"filename",
     (getter)nb_DataTrack_getfilename, (setter)nb_DataTrack_setfilename,
     "The track filename, it should point to a ISO 9660 file.",
     NULL},
    {NULL}  /* Sentinel */
};

static PyTypeObject nb_AudioTrack_Type = {
    PyObject_HEAD_INIT(NULL)
    0,                         /*ob_size*/
    "cajaburn.AudioTrack", /*tp_name*/
    sizeof(nb_Track),          /*tp_basicsize*/
    0,                         /*tp_itemsize*/
    (destructor)nb_Track_dealloc,                         /*tp_dealloc*/
    0,                         /*tp_print*/
    0,                         /*tp_getattr*/
    0,                         /*tp_setattr*/
    0,                         /*tp_compare*/
    0,                         /*tp_repr*/
    0,                         /*tp_as_number*/
    0,                         /*tp_as_sequence*/
    0,                         /*tp_as_mapping*/
    0,                         /*tp_hash */
    0,                         /*tp_call*/
    0,                         /*tp_str*/
    0,                         /*tp_getattro*/
    0,                         /*tp_setattro*/
    0,                         /*tp_as_buffer*/
    Py_TPFLAGS_DEFAULT, /*tp_flags*/
    "Represents a recorder's audio track.", /* tp_doc */
    0,		               /* tp_traverse */
    0,		               /* tp_clear */
    0,		               /* tp_richcompare */
    0,		               /* tp_weaklistoffset */
    0,		               /* tp_iter */
    0,		               /* tp_iternext */
    0,                     /* tp_methods */
    0,                     /* tp_members */
    nb_AudioTrack_getseters,   /* tp_getset */
    &nb_Track_Type,            /* tp_base */
    0,                         /* tp_dict */
    0,                         /* tp_descr_get */
    0,                         /* tp_descr_set */
    0,                         /* tp_dictoffset */
    (initproc)nb_AudioTrack_init, /* tp_init */
    0,                         /* tp_alloc */
    nb_Track_new,                 /* tp_new */
};

static PyTypeObject nb_DataTrack_Type = {
    PyObject_HEAD_INIT(NULL)
    0,                         /*ob_size*/
    "cajaburn.DataTrack", /*tp_name*/
    sizeof(nb_Track),          /*tp_basicsize*/
    0,                         /*tp_itemsize*/
    (destructor)nb_Track_dealloc,                         /*tp_dealloc*/
    0,                         /*tp_print*/
    0,                         /*tp_getattr*/
    0,                         /*tp_setattr*/
    0,                         /*tp_compare*/
    0,                         /*tp_repr*/
    0,                         /*tp_as_number*/
    0,                         /*tp_as_sequence*/
    0,                         /*tp_as_mapping*/
    0,                         /*tp_hash */
    0,                         /*tp_call*/
    0,                         /*tp_str*/
    0,                         /*tp_getattro*/
    0,                         /*tp_setattro*/
    0,                         /*tp_as_buffer*/
    Py_TPFLAGS_DEFAULT, /*tp_flags*/
    "Represents a recorder's data track.", /* tp_doc */
    0,		               /* tp_traverse */
    0,		               /* tp_clear */
    0,		               /* tp_richcompare */
    0,		               /* tp_weaklistoffset */
    0,		               /* tp_iter */
    0,		               /* tp_iternext */
    0,                     /* tp_methods */
    0,                     /* tp_members */
    nb_DataTrack_getseters,   /* tp_getset */
    &nb_Track_Type,            /* tp_base */
    0,                         /* tp_dict */
    0,                         /* tp_descr_get */
    0,                         /* tp_descr_set */
    0,                         /* tp_dictoffset */
    (initproc)nb_DataTrack_init, /* tp_init */
    0,                         /* tp_alloc */
    nb_Track_new,                 /* tp_new */
};

int
nb_AudioTrack_Check (PyObject *object)
{
	return PyObject_TypeCheck (object, &nb_AudioTrack_Type);
}

int
nb_DataTrack_Check (PyObject *object)
{
	return PyObject_TypeCheck (object, &nb_DataTrack_Type);
}

int
nb_track_init (PyObject *module)
{
	if (PyType_Ready(&nb_Track_Type) < 0)
		return -1;
	if (PyType_Ready(&nb_DataTrack_Type) < 0)
		return -1;
	if (PyType_Ready(&nb_AudioTrack_Type) < 0)
		return -1;

	Py_INCREF(&nb_Track_Type);
	PyModule_AddObject(module, "Track", (PyObject *)&nb_Track_Type);
	
	nb_DataTrack_Type.tp_bases = Py_BuildValue ("(O)", &nb_Track_Type);
	Py_INCREF(&nb_DataTrack_Type);
	PyModule_AddObject(module, "DataTrack", (PyObject *)&nb_DataTrack_Type);
	
	nb_AudioTrack_Type.tp_bases = Py_BuildValue ("(O)", &nb_Track_Type);
	Py_INCREF(&nb_AudioTrack_Type);
	PyModule_AddObject(module, "AudioTrack", (PyObject *)&nb_AudioTrack_Type);
    
	return 0;
}
